common/src/main/scala/net/psforever/objects/avatar/PlayerControl.scala

```diff
@@ -51,102 +54,35 @@ class PlayerControl(player : Player) extends Actor
-        }
+      case Player.ImplantActivation(slot : Int, status : Int) =>
+        ImplantActivation(slot, status)
 
-      case Player.UninitializeImplant(slot: Int) =>
-        PlayerControl.UninitializeImplant(player, slot)
+      case Player.UninitializeImplant(slot : Int) =>
+        UninitializeImplant(slot)
 
-      case Player.ImplantInitializationStart(slot: Int) =>
-        val implantSlot = player.ImplantSlot(slot)
-        if(implantSlot.Installed.isDefined) {
-          if(implantSlot.Initialized) {
-            PlayerControl.UninitializeImplant(player, slot)
-          }
+      case Player.ImplantInitializationStart(slot : Int) =>
+        ImplantInitializationStart(slot)
 
-          // Start client side initialization timer
-          player.Zone.AvatarEvents ! AvatarServiceMessage(player.Name, AvatarAction.SendResponse(player.GUID, ActionProgressMessage(slot + 6, 0)))
-
-          // Callback after initialization timer to complete initialization
-          implantSlot.InitializeTimer = context.system.scheduler.scheduleOnce(implantSlot.MaxTimer seconds, self, Player.ImplantInitializationComplete(slot))
-        }
-
-      case Player.ImplantInitializationComplete(slot: Int) =>
-        val implantSlot = player.ImplantSlot(slot)
-        if(implantSlot.Installed.isDefined) {
-          player.Zone.AvatarEvents ! AvatarServiceMessage(player.Name, AvatarAction.SendResponse(player.GUID, AvatarImplantMessage(player.GUID, ImplantAction.Initialization, slot, 1)))
-          implantSlot.Initialized = true
-          if(implantSlot.InitializeTimer != Default.Cancellable) {
-            implantSlot.InitializeTimer.cancel()
-            implantSlot.InitializeTimer = Default.Cancellable
-          }
-        }
+      case Player.ImplantInitializationComplete(slot : Int) =>
+        ImplantInitializationComplete(slot)
 
       case Player.DrainStamina(amount : Int) =>
-        player.Stamina -= amount
-
-      case Player.StaminaChanged(currentStamina : Int) =>
-        if(currentStamina == 0) {
-          player.Fatigued = true
-          player.skipStaminaRegenForTurns += 4
-          player.Implants.indices.foreach { slot => // Disable all implants
-            self ! Player.ImplantActivation(slot, 0)
-            player.Zone.AvatarEvents ! AvatarServiceMessage(player.Zone.Id, AvatarAction.SendResponseTargeted(player.GUID, AvatarImplantMessage(player.GUID, ImplantAction.OutOfStamina, slot, 1)))
-          }
-        } else if (player.Fatigued && currentStamina >= 20) {
-          player.Fatigued = false
-          player.Implants.indices.foreach { slot => // Re-enable all implants
-            player.Zone.AvatarEvents ! AvatarServiceMessage(player.Zone.Id, AvatarAction.SendResponseTargeted(player.GUID, AvatarImplantMessage(player.GUID, ImplantAction.OutOfStamina, slot, 0)))
-          }
+        player.Stamina += amount
+        UpdateStamina()
+
+      case Player.StaminaChanged(changeInStamina : Int) =>
```

Comment:
Do we need both `DrainStamina` and `StaminaChanged` when they seem to do the same thing, except one has a check that the stamina has actually changed before calling `UpdateStamina`?

Reply:
If you want, I can exchange `DrainStamina(Int)` into a plain `UpdateStamina()`.

Reply:
Would replacing `Player.DrainStamina` calls with `Player.StaminaChanged` not do essentially the exact same thing?

Reply:
The calls to `DrainStamina` would still have to be changed to `StaminaChanged`, and with an negated value.  But, when it comes to the line `target.Actor ! Player.DrainStamina(0)` being changed to `ChangeStamina` ...
```
val beforeStamina = player.Stamina //50
val afterStamina = player.Stamina += 0 //50 + 0 = 50
if(beforeStamina != afterStamina) { //50 != 50, false
  UpdateStamina() //doesn't get called
}
```

Reply:
Right I see what you're saying now, I guess yeah, renaming `DrainStamina` to just call `UpdateStamina` makes more sense, especially since it's not really being used to drain stamina at all, just trigger an update.

Reply:
As an extra thought: Alternatively an optional boolean on `Player.StaminaChanged` to always send an update